src/elfpy/pricing_models.py

```diff
@@ -539,6 +542,180 @@ class HyperdrivePricingModel(PricingModel):
+        d_token_reserves = (share_reserves + d_share_reserves) / 2 * (
+            init_share_price * (1 + rate * time_remaining) ** (1 / stretched_time_remaining) - share_price
+        ) - bond_reserves
+        return lp_in, base_asset_out, d_token_reserves
+
+    def calc_tokens_out_given_lp_in(
+        self,
+        lp_in,
+        share_reserves,
+        bond_reserves,
+        share_buffer,
+        init_share_price,
+        share_price,
+        liquidity_pool,
+        rate,
+        time_remaining,
+        stretched_time_remaining,
+    ):
+        assert (
+            lp_in > 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected lp_in > 0, not {lp_in}!"
+        assert (
+            share_reserves > 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected share_reserves > 0, not {share_reserves}!"
+        assert (
+            bond_reserves >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected bond_reserves >= 0, not {bond_reserves}!"
+        assert (
+            share_buffer >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected share_buffer >= 0, not {share_buffer}!"
+        assert (
+            liquidity_pool >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected liquidity_pool >= 0, not {liquidity_pool}!"
+        assert (
+            rate >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected rate >= 0, not {rate}!"
+        assert (
+            1 > time_remaining >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected 1 > time_remaining >= 0, not {time_remaining}!"
+        assert (
+            stretched_time_remaining >= 0
+        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected stretched_time_remaining >= 0, not {stretched_time_remaining}!"
+        assert share_price >= init_share_price >= 1, (
+            "pricing_models.calc_lp_out_given_tokens_in: ERROR: expected share_price >= init_share_price >= 1, not"
+        )
+        if self.verbose:
+            print(f"  inputs: lp_in={lp_in}, share_reserves={share_reserves}, bond_reserves={bond_reserves}, share_buffer={share_buffer}, init_share_price={init_share_price}, share_price={share_price}, liquidity_pool={liquidity_pool}, rate={rate}, time_remaining={time_remaining}, stretched_time_remaining={stretched_time_remaining}")
+        d_base_reserves = share_price * (share_reserves - share_buffer) * lp_in / liquidity_pool
+        d_share_reserves = d_base_reserves / share_price
+        if self.verbose:
+            print(f"  d_share_reserves={d_share_reserves} (d_base_reserves / share_price = {d_base_reserves} / {share_price})")
+        d_token_reserves = (share_reserves + d_share_reserves) / 2 * (
+            init_share_price * (1 + rate * time_remaining) ** (1 / stretched_time_remaining) - share_price
+        ) - bond_reserves
```

Comment:
Couple things:

1. `d_token_reserves` => `d_bonds`.
2. `(share_reserves + d_shares) / 2` should be `(share_reserves - d_shares) / 2` since the shares are leaving the system.
3. `- bond_reserves` must be within the parenthesis. 

Finally (and this applies everywhere in this file), I'd be more comfortable with putting parenthesis around things like `(share_reserves + d_shares) / 2` so that we don't accidentally run into an order of operations problem.